packages/flutter/lib/src/material/navigation_rail.dart

```diff
@@ -0,0 +1,849 @@
+  ///
+  /// If [backgroundColor], [elevation], [groupAlignment], [labelType],
+  /// [unselectedLabelTextStyle], [unselectedLabelTextStyle],
+  /// [unselectedIconTheme], or [selectedIconTheme] are null, then their
+  /// [NavigationRailThemeData] values will be used. If the corresponding
+  /// [NavigationRailThemeData] property is null, then the navigation rail
+  /// defaults are used. See the individual properties for more information.
+  ///
+  /// Typically used within a [Row] of the [Scaffold.body] property.
+  const NavigationRail({
+    this.backgroundColor,
+    this.extended = false,
+    this.leading,
+    this.trailing,
+    @required this.destinations,
+    this.selectedIndex = 0,
+    this.onDestinationSelected,
+    this.elevation,
+    this.groupAlignment,
+    this.labelType,
+    this.unselectedLabelTextStyle,
+    this.selectedLabelTextStyle,
+    this.unselectedIconTheme,
+    this.selectedIconTheme,
+    this.minWidth = _minRailWidth,
+    this.minExtendedWidth = _minExtendedRailWidth,
+  }) :  assert(destinations != null && destinations.length >= 2),
+        assert(0 <= selectedIndex && selectedIndex < destinations.length),
+        assert(elevation == null || elevation > 0),
+        assert(minWidth == null || minWidth > 0),
+        assert(minExtendedWidth == null || minExtendedWidth > 0),
+        assert((minWidth == null || minExtendedWidth == null) || minExtendedWidth >= minWidth),
+        assert(extended != null),
+        assert(!extended || (labelType == null || labelType == NavigationRailLabelType.none));
+
+  /// Sets the color of the Container that holds all of the [NavigationRail]'s
+  /// contents.
+  ///
+  /// The default value is [NavigationRailThemeData.backgroundColor]. If
+  /// [NavigationRailThemeData.backgroundColor] is null, then the default value
+  /// is [ThemeData.colorScheme.surface].
+  final Color backgroundColor;
+
+  /// Indicates that the [NavigationRail] should be in the extended state.
+  ///
+  /// The extended state has a wider rail container, and the labels are
+  /// positioned next to the icons.
+  ///
+  /// The rail will implicitly animate between the extended and normal state.
+  ///
+  /// If the rail is going to be in the extended state, then the [labelType]
+  /// must be set to [NavigationRailLabelType.none].
+  ///
+  /// The default value is false.
+  final bool extended;
+
+  /// The leading widget in the rail that is placed above the destinations.
+  ///
+  /// The leading widget is placed at the top of the rail. It's location is not
+  /// affected by [groupAlignment]
+  ///
+  /// This is commonly a [FloatingActionButton], but may also be a non-button,
+  /// such as a logo.
+  ///
+  /// By default, it is null, and no leading widget is built.
+  final Widget leading;
+
+  /// The trailing widget in the rail that is placed below the destinations.
+  ///
+  /// The trailing widget is placed below the last [NavigationRailDestination].
```

Comment:
I think it needs to be a slot for an extra widget, because destinations are not configured by passing in a widget or list of widgets, but rather a data model that creates a widget. Leaving it open like this also makes it easy to place any horizontal dividers.